include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -815,19 +817,46 @@ struct __simple_write_to_idx
 };
 
 template <typename _Predicate>
-struct __gen_count_pred
+struct __gen_mask
+{
+    template <typename _InRng>
+    bool
+    operator()(_InRng&& __in_rng, std::size_t __idx) const
+    {
+        return __pred(__in_rng[__idx]);
+    }
+    _Predicate __pred;
+};
+
+template <typename _BinaryPredicate>
+struct __gen_unique_mask
+{
+    template <typename _InRng>
+    bool
+    operator()(_InRng&& __in_rng, std::size_t __idx) const
+    {
+        if (__idx == 0)
```

Comment:
Are there any tricks we could use to remove this conditional entirely? It feels like this is a lot of overhead for special handling of the first element.

Reply:
This is a good question...here's my best idea:
Its a little ugly, but it will be faster than the "cleaner" options.

Add a template bool to the `parallel_scan_copy` (and turn it on for `pattern_unique_copy`) which special cases the first element in the second kernel, and adds an offset of 1 to the scan operation. This would allow us to just have a single operation for the 0th index work item to make the copy, and then scan without the branch starting at element 1 instead of 0 for both kernels.  

The cleaner option would be a separate memcpy which handles the first element (and offsetting the start of the list), but if we handled that at a high level, it would be a separate memcpy or kernel which I think ends up slower.

What do you think?

Reply:
This has been applied via #1743 (merged).